rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,322 @@
+
+Templates will now also have a status with a `Ready` condition which will be `True` if the referenced `ClusterResourceStatusRule` exists (or if no rule is referenced), and `False` otherwise. Blueprints will now wait to report that they are `Ready` until their referenced templates are also `Ready`.
+
+Each owner will display a new condition in the status called `ResourcesSuceeded` which will only be updated to `True` when all resources in the blueprint have succeeded. Cartographer will not update the top level state of an owner to `Ready: True` until `ResourcesSucceeded` is `True`.
+
+# Migration
+[migration]: #migration
+
+Given that Cartographer will default to `Succeeded` once the resource is applied and outputs are read, this will not be a breaking change.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+- The owner status will potentially grow significantly, based on how many matchers are in a `ClusterResourceStatusRule`.
+- We only allow for reflecting three states, though this design allows for creating more states in the future.
+- This design does not allow authors to define the states themselves.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+  - The `ClusterResourceStatusRule` could instead be inlined in the templates.
+    - Separating these rules in a separate CRD allows for re-usability of those rules.
+  - We could also not require the templates to specify the `ClusterResourceStatusRule` and instead have those rules automatically select templates based on the `kind` and `apiVersion` of the resource they are stamping out.
+    - This could be a bit confusing because of ytt. A template could end up stamping out different types of resources and so Cartographer would only be able to inform users which `ClusterResourceStatusRule` is used for a given template after an owner is applied, and it could then be reflected in the owner status.
+- What is the impact of not doing this?
+  - Users/UI creators will have to have in depth knowledge of the underlying resources to determine the health of the owner.
+
+# Prior Art
+[prior-art]: #prior-art
+
+N/A
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should the two states be `Succeeded` and `Failed`?
+- Is it right for Cartographer to default to `Succeeded`?
+
+# Spec. Changes
+[spec-changes]: #spec-changes
+- All templates will now have an optional `resourceStatusRuleRef` which will take a `name`. In the future, it could also take a type and namespace if we introduce namespace-scoped resourceStatusRules.
+- Owner `status.resources` will now have `stampedStatus` on each resource.
+- A new CRD, `ClusterResourceStatusRule` will be created with the following spec:
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterResourceStatusRule
+metadata: {}
+spec:
+
+  succeeded:
```

Comment:
I'd be open to adding 'unknown' as a field, but then how would we distinguish between `unknown` as defined by the ClusterResourceStatusRule, and unknown where none of the matchers have resolved and cartographer truly does not know the state?

Reply:
So something like an extra field for `unknownMessagePath`? Or `uknown.messagePath` in case we want to add more later?

Reply:
I think it makes sense to have the same syntax (matchFields, matchConditions) for every possible state, so `Unknown` wouldn't be any different. It would mean that int the workload status, sometimes `Unknown` has a corresponding `fields` or `status` entry, and sometimes not.

Reply:
That is something we could easily add to the spec if we received feedback on it. If we are adding it now, it really becomes 4 states and we are collapsing 2.

Reply:
To clarify this discussion in light of recent changes the the RFC, if you specified `singleCondition` then if that condition was `Unknown`, you'd still get the message and reason from that condition propagated up to the owner.

So this conversation seems only relevant for the case of `multiMatch`. I think @emmjohnson's suggestion that we later add `unknown` if we receive feedback requesting it makes a lot of sense. Interested to hear what others think, though.